Bluetooth: Audio: Upstream media control

asbjornsabo opened the pull request:
This PR upstreams the media control implementation from the topic-le-audio branch.
This includes
- media control service
- dummy media player that acts as a back end providing functionality for the media control service
- media control client
- temporary object transfer client implementation
  (to be cleaned up and moved alongside the object transfer service, but required for media control)
- media control proxy, which unifies access to local media players (direct access) and remote media players (access using the media control client).
- build tests
- BabbleSim tests for all functionality
- Shell commands to exercise the functionality

The code is quite mature, but changes are expected.  The main purpose now is to get the code upstream in its current state, to avoid having the topic branch. 

This upstreaming PR introduces no changes to the functionality of the code, that is kept as it was in the topic branch.
The only changes are minor clean-ups and adaptions to new locations - these changes are listed in each of the commits.

asbjornsabo committed:
Bluetooth: Audio: Media control client

This commit adds the Media Control Client from the topic-le-audio
branch.  This is a part of the upmerge of the le-audio media control
files.

This client has been developed and maintained over a couple of years
now, and is mature.  During the development it has passed both
IOP-testing and PTS qualification testing.

The commit is a pure copy of the files and content in the
topic-le-audio branch, with the following exceptions:

- files are in bluetooth/audio instead of bluetooth/host/audio, with
  some include paths updated as a consequence
- as a consequence, CMake files and Kconfig files updates are done in
  other locations
- a macro for debug output of Object ID values has been (temporarily)
  added to mcc.h, to avoid a dependency
- a blank line added after a declaration and an overlong line split,
  to pass check_compliance
- copyrights have been updated

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Temporary Object Transfer client

This commit adds the Object Transfer client from the topic-le-audio
branch.  This is a part of the upmerge of the le-audio media control
files.  Media control depends upon Object Transfer for feature
completeness.

This client works, and has been used for IOP- and PTS-testing of the
media control server and client.  It does, however, need an overhaul
before being made "Zephyr official", so for now it will be kept as an
internal file to the le-audio code.

The commit is a pure copy of the files and content in the
topic-le-audio branch, with the following exceptions:

- files are in bluetooth/audio instead of bluetooth/host/audio, with
  an include path updated as a consequence
- a sa consequence, CMake files and Kconfig files updates are done in
  other locations
- In otc.c a superfluous semi-colon has been removed to please
  check_compliance
- Kconfig.ots has been renamed to Kconfig.otc, as there is only
  otc-content left in that file
- copyrights have been updated
- add TODO that these files are to be cleaned up and moved

The Object Transfer Client in this commit was originally written by
Leif Aschehoug <Leif-Alexandre.Aschehoug@nordicsemi.no>.
It has later been updated and maintained by
Emil Gydesen <emil.gydesen@nordicsemi.no> and
Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

ots

asbjornsabo committed:
Bluetooth: Audio: UUID values for media control

UUID values for the Media Control Service - service and
characteristics.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media control service

This commit adds the Media Control Service, and a dummy media player,
from the topic-le-audio branch.  This is a part of the upmerge of the
le-audio media control files.

This service has been developed and maintained over a couple of years
now, and is mature.  During the development it has passed both
IOP-testing and PTS qualification testing.

The commit is a pure copy of the files and content in the
topic-le-audio branch, with the following exceptions:

- files are in bluetooth/audio instead of bluetooth/host/audio, with
  some include paths updated as a consequence
- as a consequence, CMake files and Kconfig files updates are done in
  other locations
- copyrights have been updated

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media proxy

This commit adds the media proxy from the topic-le-audio branch.
This is a part of the upmerge of the le-audio media control files.

The media proxy adds a common API to register and access local media
players, and to discover and access remote media players over
bluetooth using the media control client and a remote media control
service.

The commit is a pure copy of the files and content in the
topic-le-audio branch, with the following exceptions:

- files are in bluetooth/audio instead of bluetooth/host/audio, with
  some include paths updated as a consequence
- as a consequence, CMake files and Kconfig files updates are done in
  other locations

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media control project settings

Upmerge of the project settings for media control from the
topic-le-audio branch.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

audio.conf changes

Thalley commented on subsys/bluetooth/audio/Kconfig.mcs:
```diff
@@ -0,0 +1,210 @@
+endif # BT_MCC
+
+
+if BT_MCS || BT_MCC
+
+config BT_MCS_MEDIA_PLAYER_NAME_MAX
```
Are the MCC still relying on Kconfigs named BT_MCS? That's somewhat surprising and I guess should also be fixed at some point.

asbjornsabo replied:
Agree - noted

Thalley commented on subsys/bluetooth/audio/Kconfig.otc:
```diff
@@ -0,0 +1,34 @@
+# Bluetooth Audio - Object transfer configuration options
```
I guess why this was moved to the `audio` directory, but can't we just as easily put it next to OTS?

Thalley replied:
This comment applies for the entirety of OTC (header filers, Kconfigs, etc.)

asbjornsabo replied:
The goal is to put it there.  But I would like to keep the OTC "hidden" here until we have cleaned it up and brought it up to the standard of the OTS.

Thalley commented on subsys/bluetooth/audio/Kconfig.otc:
```diff
@@ -0,0 +1,34 @@
+
+config BT_DEBUG_OTC
+	bool "Object Transfer Client debug"
+	depends on BT_AUDIO_DEBUG
```
Should not depend on BT_AUDIO_DEBUG

asbjornsabo replied:
I see your point, I think.  As a general service/client, it should not depend upon anything AUDIO.
Added a TODO to remove this when OTC is moved out of the audio directory

Thalley commented on subsys/bluetooth/audio/mcs.c:
```diff
@@ -0,0 +1,989 @@
+/** @file
+ *  @brief Bluetooth Media Control Service - d09r01 IOP1 2019-08-22
```
Isn't it version 1.0?

asbjornsabo replied:
It is.  Removed the whole version string - the spec is stable now, less need to track which version.

Thalley commented on subsys/bluetooth/audio/mcs_internal.h:
```diff
@@ -0,0 +1,25 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_HOST_AUDIO_MTS_INTERNAL_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_HOST_AUDIO_MTS_INTERNAL_H_
+
+#include <stdbool.h>
+#include <zephyr/types.h>
+#include <bluetooth/gatt.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+int bt_mcs_init(struct bt_ots_cb *ots_cbs);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* ZEPHYR_INCLUDE_BLUETOOTH_HOST_AUDIO_OTS_INTERNAL_H_ */
```
```suggestion
#ifndef ZEPHYR_SUBSYS_BLUETOOTH_AUDIO_OTS_INTERNAL_H_
#define ZEPHYR_SUBSYS_BLUETOOTH_AUDIO_OTS_INTERNAL_H_

#include <stdbool.h>
#include <zephyr/types.h>
#include <bluetooth/gatt.h>

#ifdef __cplusplus
extern "C" {
#endif

int bt_mcs_init(struct bt_ots_cb *ots_cbs);

#ifdef __cplusplus
}
#endif

#endif /* ZEPHYR_SUBSYS_BLUETOOTH_AUDIO_OTS_INTERNAL_H_ */
```

I see that some of the other services (e.g. VCS and MICS have the same issue).

asbjornsabo replied:
Fixed.
(It is MCS, though, not OTS.  And the whole init function should be removed, I think.)

Thalley commented on subsys/bluetooth/audio/media_proxy_internal.h:
```diff
@@ -0,0 +1,129 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_AUDIO_MEDIA_PROXY_INTERNAL_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_AUDIO_MEDIA_PROXY_INTERNAL_H_
```
Ditto

Thalley commented on subsys/bluetooth/audio/mpl.c:
```diff
@@ -0,0 +1,2923 @@
+/*  Media player skeleton implementation */
+
+/*
+ * Copyright (c) 2019 - 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Please tell me if this is an unreasonable request for this PR: Since the Media Player is not BT specific, and in fact probably shouldn't even reside in `zephyr/subsystem`, would it require much to move this to a sample instead, which it technically is?

asbjornsabo replied:
I would prefer to do that as a separate PR (if it is to be done).
But I am also not sure that would be wise - it would make the stack depend upon a sample, which seems the wrong way around.

Thalley replied:
The stack doesn't really depend on the media player, does it? Sure, you can't use the media proxy without it, but it should still build just fine. 

As I see it at least, both the media player and the media controller could reside outside of the Zephyr Subsystem/Stack, but let's continue this outside the scope of this PR :)

Thalley commented on subsys/bluetooth/audio/mpl_internal.h:
```diff
@@ -0,0 +1,153 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_AUDIO_MPL_INTERNAL_
+#define ZEPHYR_INCLUDE_BLUETOOTH_AUDIO_MPL_INTERNAL_
```
Ditto

Thalley commented on subsys/bluetooth/audio/otc_internal.h:
```diff
@@ -0,0 +1,36 @@
+/* @file
+ * @brief Object Transfer client internal header file
+ *
+ * For use with the Object Transfer Service Client (OTC)
+ *
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Given that this file only implements a single macro (which I suppose may as well be defined in the public otc.h or even ots.h files) that is only used a single place, maybe this file is overkill to have.

Thalley commented on subsys/bluetooth/audio/ots.h:
```diff
@@ -0,0 +1,322 @@
+/* @file
+ * @brief Object Transfer
+ *
+ * For use with the Object Transfer Service (OTS)
+ *
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_HOST_AUDIO_OTS_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_HOST_AUDIO_OTS_H_
```
Oh, I was under the impression that we actually had started using the upstreamed OTS implementation, but I see that the otc implementation still uses this. If we don't remove the otc_internal, suggest to move this code into that for now, with a TODO.

Thalley commented on subsys/bluetooth/audio/ots_internal.h:
```diff
@@ -0,0 +1,48 @@
+/* @file
+ * @brief Object Transfer internal header file
+ *
+ * For use with the Object Transfer Service (OTS)
+ *
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Same comment for this as for the ots.h file.

Thalley commented on tests/bluetooth/shell/audio.conf:
```diff
@@ -54,3 +54,28 @@ CONFIG_BT_MICS=y
 CONFIG_BT_MICS_CLIENT=y
 CONFIG_BT_MICS_CLIENT_MAX_AICS_INST=1
+
+CONFIG_BT_MCS=y
+CONFIG_BT_MCS_MEDIA_PLAYER_NAME_MAX=30
+CONFIG_BT_MCS_MEDIA_PLAYER_NAME="My media player"
+CONFIG_BT_MCS_ICON_URL_MAX=27
+CONFIG_BT_MCS_ICON_URL="http://www.somecompany.com"
+CONFIG_BT_MCS_TRACK_TITLE_MAX=40
+CONFIG_BT_MCS_GROUP_TITLE_MAX=40
+CONFIG_BT_MCS_SEGMENT_NAME_MAX=35
+CONFIG_BT_MCS_ICON_BITMAP_SIZE=321
+CONFIG_BT_MCC=y
+CONFIG_BT_MCC_OTS=y
+CONFIG_BT_MCC_TOTAL_OBJ_CONTENT_MEM=4096
+# TODO Check which value is sensible in the line below
+# Must be larger than any of the object sizes - icon, track, group, segments, ...
+CONFIG_BT_MCS_MAX_OBJ_SIZE=600
+CONFIG_BT_MCS_TRACK_MAX_SIZE=50
+CONFIG_BT_MCS_TRACK_SEG_MAX_SIZE=500
+CONFIG_BT_MCS_GROUP_MAX_SIZE=552
+
+# Object Transfer
+CONFIG_BT_OTS=y
+CONFIG_BT_OTS_SECONDARY_SVC=y
+CONFIG_BT_OTS_MAX_OBJ_CNT=50
+CONFIG_BT_OTC=y
```
Hmm, we enable MCS in the shell, but we don't add the MCS module? That seems odd. I guess it is just for building purposes?

asbjornsabo replied:
You commented that the BabbleSim tests should have been added, and I think you are right.
When doing that, I will also add the shells.

Thalley reviewed (changes requested):
I see that the Babblesim tests didn't make it into this PR, which I think is a shame, as we then don't run bsim tests on this PR.

asbjornsabo commented on include/bluetooth/audio/mcc.h:
```diff
@@ -0,0 +1,938 @@
+ */
+
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
Yes.  Done everywhere now, I believe.

asbjornsabo commented on include/bluetooth/audio/mcs.h:
```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2019 - 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
I am not sure yet what is best.
I can certainly move this to be internal for now - I expect it is easier to to add to the public interface later than to remove from it.
But for consistency, should the same also apply for the client header (mcc.h)?

The file is marked as experimental, so I believe it can also be left as is for now, and moved when we have decided on the API.
What do you prefer?

Thalley replied:
As long as media_proxy.h relies on this, this should remain in the public structure, but I don't think it is the intention that that media_proxy.h relies on this, right? 

Regarding if this should apply to mcc.h, that depends on whether we only want to expose the MCS and MCC functionality via the Media Proxy, or if we want to allow users to use those implementations directly.

asbjornsabo commented on subsys/bluetooth/audio/CMakeLists.txt:
```diff
@@ -21,5 +21,11 @@ if (CONFIG_BT_MICS OR CONFIG_BT_MICS_CLIENT)
+zephyr_library_sources_ifdef(CONFIG_BT_OTC otc.c)
+zephyr_library_sources_ifdef(CONFIG_BT_CCID ccid.c)
+zephyr_library_sources_ifdef(CONFIG_BT_MCC mcc.c)
+zephyr_library_sources_ifdef(CONFIG_BT_MCS mcs.c)
+zephyr_library_sources_ifdef(CONFIG_BT_MCS mpl.c)
+zephyr_library_sources_ifdef(CONFIG_BT_MCS media_proxy.c)
```
The Kconfigs should be rethought, indeed.
If there is no MCS, there is still a remote player, and media_proxy may still be used.

asbjornsabo commented on include/bluetooth/audio/media_proxy.h:
```diff
@@ -0,0 +1,1614 @@
+#include <zephyr/types.h>
+#include <bluetooth/bluetooth.h>
+
+#include "mcs.h"
```
Added TODO.  (Have not moved the mcs header file yet, though.)

asbjornsabo committed:
Bluetooth: Audio: Add build tests for media control

Add entries for media control builds to testcase list.  This is a part
of the upmerge of media control from the topic-le-audio branch.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Media control BabbleSim tests

This commit add the BabbleSim tests for the media control client, the
media control service and the media proxy.  This is a part of the
upmerge of the le-audio media control files.

These tests have been developed and maintained along with the media
control functionality.  The tests are believed to cover all
implemented media control functionality.

The commit is a pure copy of the files and content in the
topic-le-audio branch, with the exception that some include paths have
been modified to correspond to new locations of files and that the
tests have been guarded with ifdefs.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

test guards

asbjornsabo committed:
Bluetooth: Audio: Media control shell commands

This commit adds shell commands for the media control client, the
media player and the media proxy.  This is a part of the upmerge of
the le-audio media control files.

The commit is a pure copy of the files and content in the
topic-le-audio branch, with these exception:

- some include paths have been modified to correspond to new locations
  of files.
- some instances of the shell variable has been renamed to not be the
  same as the name of the shell struct tag.
- spaces have been replaced with TABs a couple of places
- remove cplusplus guards
- changed a debug dependency for MCC from MCS to MCC
- changed log name for the media controller shell

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

mcc.c shell - replace name of shell variable

tmp

f

shell

asbjornsabo committed:
Bluetooth: Audio: Bugfix - add missing parameters

Add missing parameter to shell_print call

(The bug exists in the topic-le-audio branch also, but was not
discovered there.  In the upmerge branch, it causes build failure.)

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Thalley reviewed (changes requested):
Minor comments

asbjornsabo commented:
The Scancode license check is failing due to a false positive, reported in 
https://github.com/nexB/scancode-toolkit/issues/2304
A fix has supposedly been committed here:
https://github.com/nexB/scancode-toolkit/commit/6abbc4a22973f40ab74f6f8d948dd06416c97bd4

Thalley commented on subsys/bluetooth/shell/media_controller.c:
```diff
@@ -0,0 +1,885 @@
+#include "../audio/media_proxy_internal.h" /* For MPL_NO_TRACK_ID - TODO: Fix */
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_MCS)
+#define LOG_MODULE_NAME bt_media_shell
```
```suggestion
#define LOG_MODULE_NAME bt_media_controller_shell
```

Thalley commented on subsys/bluetooth/shell/mpl.c:
```diff
@@ -0,0 +1,264 @@
+#define LOG_MODULE_NAME bt_mpl_shell
+#include "common/log.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```
Ditto

asbjornsabo committed:
Bluetooth: Audio: Workaround for false license positive

Add a workaround to stop Scancode from falsely detecting an MPL
license in the code.

See https://github.com/nexB/scancode-toolkit/issues/2304

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on subsys/bluetooth/audio/ots.h:
```diff
@@ -0,0 +1,322 @@
+/* @file
+ * @brief Object Transfer
+ *
+ * For use with the Object Transfer Service (OTS)
+ *
+ * Copyright (c) 2020 - 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
TODOs seem like a good idea - added

asbjornsabo commented on subsys/bluetooth/shell/mcc.c:
```diff
@@ -0,0 +1,1539 @@
+#define LOG_MODULE_NAME bt_mcc_shell
+#include "common/log.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```
Only there for "hysterical raisins", I guess.  Removed.

Thalley reviewed (approved):
All my comments have been addressed
